Fix pty support for "env" in jobstart/termopen

jamessan opened the pull request:
Pulling together the work that @teto and @erw7 started (c.f., #11839).

Closes #11751

---
- [x] Fix memory leak
- [x] Convert env block to `wchar_t` on Windows
  - [x] Pass environment through when creating pty job on Windows (`pty_process_spawn`)
- [x] Iterate through environment in reverse in `f_environ` so dict has relevant values if there are duplicates
  - [x] ~Determine how to handle de-duplication of case-insensitive variable names (yay, Windows)~ going to punt on this for now

jamessan committed:
Use dict_T to pass env vars to process spawning code

Co-authored-by: Matthieu Coudron <mattator@gmail.com>

jamessan committed:
Support specifying "env" option for termopen()

Co-authored-by: erw7 <erw7.github@gmail.com>

jamessan committed:
Common handling of required/ignored env vars

When starting a pty job, there are certain env vars that we need to
either add or remove.  Currently, there are two relevant scenarios.

* Removing irrelevant env vars on Unix, mostly related to the terminal
  hosting nvim since they do not apply to a libvterm-hosted terminal.

* Adding required env vars for Windows jobs.

jamessan committed:
eval/environ: Prefer the last definition of an env var

It's possible for the environment variable block given to nvim to
contain multiple definitions for the same env var.  In this case, nvim
should preserve the last one defined.

jamessan committed:
Pass environment on to pty processes on Windows

vim-patch:8.2.0239: MS-Windows: 'env' job option does not override existing vars

Problem:  MS-Windows: 'env' job option does not override existing
          environment variables.  (Tim Pope)
Solution: Set the environment variables later. (Yasuhiro Matsumoto,
          closes vim/vim#5485, closes vim/vim#5608)
https://github.com/vim/vim/commit/355757aed6ae2ae5446882570d89f243e4805937

Co-authored-by: erw7 <erw7.github@gmail.com>

jamessan committed:
test(job): Ensure job-specific env var overrides global env var

jamessan commented:
@erw7 Any chance you would be able to look into the Windows failures?  I don't have an environment I can use to debug.

erw7 commented:
I think you need to change 665c9c11bbbf770ba81f6296dcb076c773f1754a to the following.

```diff
diff --git a/src/nvim/os/pty_process_win.c b/src/nvim/os/pty_process_win.c
index 9d42de377..fd3f3159e 100644
--- a/src/nvim/os/pty_process_win.c
+++ b/src/nvim/os/pty_process_win.c
@@ -505,14 +505,16 @@ static int build_env_block(dict_T *denv, wchar_t **env_block)
   env_block_len++;
 
   *env_block = xmalloc(sizeof(**env_block) * env_block_len);
+  wchar_t *pos = *env_block;
 
   QUEUE_FOREACH(q, &env_q) {
     EnvNode *env_node = QUEUE_DATA(q, EnvNode, node);
-    memcpy(env_block, env_node->str, env_node->len);
-    env_block += env_node->len;
+    memcpy(pos, env_node->str, env_node->len * sizeof(*pos));
+    pos += env_node->len;
+    *pos++= L'\0';
   }
 
-  *env_block = '\0';
+  *pos = L'\0';
 
 cleanup:
   q = QUEUE_HEAD(&env_q);
```

erw7 commented on src/nvim/os/pty_process_win.c:
```diff
@@ -454,3 +465,68 @@ int translate_winpty_error(int winpty_errno)
+    memcpy(pos, env_node->str, env_node->len * sizeof(*pos));
+    pos += env_node->len;
+    *pos = L'\0';
+    pos++;
```
If `env_node->len` is set to `+1` from the beginning, there is no need to assign a terminator (`L'\0'`). Therefore, lines 514-515 need to be deleted.

erw7 commented:
The reason why the test fails is because ANSI escape sequences are included in the output. By making the following changes, the test now passes.

CI: https://github.com/erw7/neovim/actions/runs/482401432
Ref: #11726

```diff
diff --git a/test/functional/core/job_spec.lua b/test/functional/core/job_spec.lua
index d71dd9aaa..6b9ba5a07 100644
--- a/test/functional/core/job_spec.lua
+++ b/test/functional/core/job_spec.lua
@@ -33,7 +33,7 @@ describe('jobs', function()
     function! Normalize(data) abort
       " Windows: remove ^M
       return type([]) == type(a:data)
-        \ ? map(a:data, 'substitute(v:val, "\r", "", "g")')
+        \ ? map(a:data, 'substitute(substitute(v:val, "\r", "", "g"), "\x1b\\%(\]\\d\\+;[^\x07]*\x07\\|\[[\x30-\x3F]*[\x40-\x7E]\\)", "", "g")')
         \ : a:data
     endfunction
     function! OnEvent(id, data, event) dict
```

clason commented:
I cannot reproduce the CI test failures on macOS (Intel 11.2) locally, but I do get two failing tests:

* the first looks directly related (as it's the new test):
```
[  ERROR   ] test/functional/core/job_spec.lua @ 79: jobs append environment with pty #env
test/functional/helpers.lua:197:
==============================================================================
test/helpers.lua:73: Expected objects to be the same.
Passed in:
(table: 0x04e60440) {
 *[1] = {
    [1] = 'notification'
    [2] = 'stdout'
   *[3] = {
      [1] = 0
     *[2] = {
       *[1] = 'goodbye world abc'
        [2] = '' } } } }
Expected:
(table: 0x04f060d0) {
 *[1] = {
    [1] = 'notification'
    [2] = 'stdout'
   *[3] = {
      [1] = 0
     *[2] = {
       *[1] = 'hello world abc'
        [2] = '' } } } }
------------------------------------------------------------------------------
$NVIM_LOG_FILE: /Users/clason/build/neovim/build/.nvimlog
(last 10 lines)
INFO  2021-01-30T18:20:06.653 11844 set_init_2:754: startup runtimepath/packpath value: /Users/clason/build/neovim/build/Xtest_xdg/config/nvim,/etc/xdg/nvim,/Users/clason/build/neovim/build/Xtest_xdg/share/nvim/site,/usr/local/share/nvim/site,/usr/share/nvim/site,/Users/clason/build/neovim/runtime,/usr/local/lib/nvim,/usr/share/nvim/site/after,/usr/local/share/nvim/site/after,/Users/clason/build/neovim/build/Xtest_xdg/share/nvim/site/after,/etc/xdg/nvim/after,/Users/clason/build/neovim/build/Xtest_xdg/config/nvim/after
INFO  2021-01-30T18:20:06.656 11844 main:559: starting main loop
INFO  2021-01-30T18:20:06.659 11844 channel_create_event:204: new channel 3 (?) : {"id": 3, "mode": "bytes", "stream": "job"}
INFO  2021-01-30T18:20:06.664 11844 on_process_exit:393: exited: pid=11845 status=0 stoptime=0
WARN  2021-01-30T18:20:06.665 11844 call_set_error:627: RPC: ch 1 was closed by the client
INFO  2021-01-30T18:20:06.666 11844 os_exit:586: Nvim exit: 0
INFO  2021-01-30T18:20:06.672 11847 set_init_2:754: startup runtimepath/packpath value: /Users/clason/build/neovim/build/Xtest_xdg/config/nvim,/etc/xdg/nvim,/Users/clason/build/neovim/build/Xtest_xdg/share/nvim/site,/usr/local/share/nvim/site,/usr/share/nvim/site,/Users/clason/build/neovim/runtime,/usr/local/lib/nvim,/usr/share/nvim/site/after,/usr/local/share/nvim/site/after,/Users/clason/build/neovim/build/Xtest_xdg/share/nvim/site/after,/etc/xdg/nvim/after,/Users/clason/build/neovim/build/Xtest_xdg/config/nvim/after
INFO  2021-01-30T18:20:06.675 11847 main:559: starting main loop
INFO  2021-01-30T18:20:06.678 11847 channel_create_event:204: new channel 3 (?) : {"id": 3, "mode": "bytes", "stream": "job", "pty": "/dev/ttys004"}
INFO  2021-01-30T18:20:06.685 11847 on_process_exit:393: exited: pid=11848 status=0 stoptime=0
------------------------------------------------------------------------------
```
(The one before *without* `pty` passes.)

* The second is less expected:

```
[  ERROR   ] test/functional/core/startup_spec.lua @ 236: startup does not crash when expanding cdpath during early_init
test/helpers.lua:73: Expected objects to be the same.
Passed in:
(nil)
Expected:
(string) ',~doesnotexist'
------------------------------------------------------------------------------
$NVIM_LOG_FILE: /Users/clason/build/neovim/build/.nvimlog
(last 10 lines)
INFO  2021-01-30T18:25:56.259 12225 set_init_2:754: startup runtimepath/packpath value: /Users/clason/build/neovim/build/Xtest_xdg/config/nvim,/etc/xdg/nvim,/Users/clason/build/neovim/build/Xtest_xdg/share/nvim/site,/usr/local/share/nvim/site,/usr/share/nvim/site,/Users/clason/build/neovim/runtime,/usr/local/lib/nvim,/usr/share/nvim/site/after,/usr/local/share/nvim/site/after,/Users/clason/build/neovim/build/Xtest_xdg/share/nvim/site/after,/etc/xdg/nvim/after,/Users/clason/build/neovim/build/Xtest_xdg/config/nvim/after
WARN  2021-01-30T18:25:56.260 12225 call_set_error:627: RPC: ch 1 was closed by the client
INFO  2021-01-30T18:25:56.260 12225 os_exit:586: Nvim exit: 0
INFO  2021-01-30T18:25:56.260 12226 set_init_2:754: startup runtimepath/packpath value: /Users/clason/build/neovim/build/Xtest_xdg/config/nvim,/etc/xdg/nvim,/Users/clason/build/neovim/build/Xtest_xdg/share/nvim/site,/usr/local/share/nvim/site,/usr/share/nvim/site,/Users/clason/build/neovim/runtime,/usr/local/lib/nvim,/usr/share/nvim/site/after,/usr/local/share/nvim/site/after,/Users/clason/build/neovim/build/Xtest_xdg/share/nvim/site/after,/etc/xdg/nvim/after,/Users/clason/build/neovim/build/Xtest_xdg/config/nvim/after
INFO  2021-01-30T18:25:56.261 12226 main:559: starting main loop
WARN  2021-01-30T18:25:56.263 12226 call_set_error:627: RPC: ch 1 was closed by the client
INFO  2021-01-30T18:25:56.264 12226 os_exit:586: Nvim exit: 0
INFO  2021-01-30T18:25:56.272 12227 set_init_2:754: startup runtimepath/packpath value: /Users/clason/build/neovim/build/Xtest_xdg/config/nvim,/etc/xdg/nvim,/Users/clason/build/neovim/build/Xtest_xdg/share/nvim/site,/usr/local/share/nvim/site,/usr/share/nvim/site,/Users/clason/build/neovim/runtime,/usr/local/lib/nvim,/usr/share/nvim/site/after,/usr/local/share/nvim/site/after,/Users/clason/build/neovim/build/Xtest_xdg/share/nvim/site/after,/etc/xdg/nvim/after,/Users/clason/build/neovim/build/Xtest_xdg/config/nvim/after
WARN  2021-01-30T18:25:56.273 12227 call_set_error:627: RPC: ch 1 was closed by the client
INFO  2021-01-30T18:25:56.274 12227 os_exit:586: Nvim exit: 0
------------------------------------------------------------------------------
```

mjlbach commented:
@jamessan This fixes one of the test failures

```patch
diff --git a/src/nvim/os/env.c b/src/nvim/os/env.c
index 008f5ef63..3540d52d9 100644
--- a/src/nvim/os/env.c
+++ b/src/nvim/os/env.c
@@ -25,7 +25,7 @@
 #include "nvim/mbyte.h"  // for utf8_to_utf16, utf16_to_utf8
 #endif
 
-#ifdef HAVE__NSGETENVIRON
+#ifdef have__nsgetenviron
 #include <crt_externs.h>
 #endif
 
@@ -204,7 +204,7 @@ size_t os_get_fullenv_size(void)
 
   FreeEnvironmentStringsW(envstrings);
 #else
-# if defined(HAVE__NSGETENVIRON)
+# if defined(have__nsgetenviron)
   char **environ = *_NSGetEnviron();
 # else
   extern char         **environ;
@@ -260,7 +260,7 @@ void os_copy_fullenv(char **env, size_t env_size)
 
   FreeEnvironmentStringsW(envstrings);
 #else
-# if defined(HAVE__NSGETENVIRON)
+# if defined(have__nsgetenviron)
   char **environ = *_NSGetEnviron();
 # else
   extern char         **environ;
@@ -320,7 +320,7 @@ char *os_getenvname_at_index(size_t index)
   FreeEnvironmentStringsW(envstrings);
   return name;
 #else
-# if defined(HAVE__NSGETENVIRON)
+# if defined(have__nsgetenviron)
   char **environ = *_NSGetEnviron();
 # else
   extern char         **environ;
diff --git a/src/nvim/os/pty_process_unix.c b/src/nvim/os/pty_process_unix.c
index 073358987..22024377b 100644
--- a/src/nvim/os/pty_process_unix.c
+++ b/src/nvim/os/pty_process_unix.c
@@ -158,7 +158,7 @@ void pty_process_teardown(Loop *loop)
 static void init_child(PtyProcess *ptyproc)
   FUNC_ATTR_NONNULL_ALL
 {
-#if defined(HAVE__NSGETENVIRON)
+#if defined(have__nsgetenviron)
   char **environ = *_NSGetEnviron();
 #else
   extern char **environ;
```

jamessan committed:
fix(pty_proc/macOS): Properly set the environment for the child

Binding _NSGetEnviron()'s return value to a local variable and then
re-binding that is incorrect.  We need to directly update what
_NSGetEnviron() refers to.

jamessan commented:
>     * the first looks directly related (as it's the new test):

Latest commit fixes this (and hopefully all the other macOS failures).

clason commented:
Both tests now pass locally for me.

clason commented:
And CI agrees (at least on macOS) \o/

jamessan committed:
Use abort() instead of assert(false) for things that should never happen

assert() is compiled out for release builds, but we don't want to
continue running in these impossible situations.

This also resolves the "implicit fallthrough" warnings for the asserts
in switch cases.

jamessan committed:
ci(sr.ht): Use RelWithDebInfo builds

Many compiler checks rely on optimizations being enabled, so we want
these to show up in CI.  Use RelWithDebInfo instead of Release so that
assert() statements aren't compiled out.

janlazo commented:
FreeBSD build cannot compile shada code because of 27a7a4d3.

jamessan commented:
> FreeBSD build cannot compile shada code because of 27a7a4d.

It compiled fine in this PR after I made that change.
